### DISCUSSION TITLE
Vulkan backend on macOS

### DISCUSSION THREAD

**Discussion by rasky** [gfx-rs/wgpu#4770] [Q&A]
Hello, I'm trying to understand whether it is possible to use the Vulkan backend on macOS. It seems I'm unable to make it work.

I've installed the Vulkan SDK on macOS (with MoltenVK). If I run `wgpu-info` it does seem that the Vulkan backend exists and is available:

```
$ cargo run --release --bin wgpu-info
    Finished release [optimized] target(s) in 0.12s
     Running `target/release/wgpu-info`
Adapter 0:
       Backend: Vulkan
          Name: "Apple M1 Pro"
      VendorID: 4203
      DeviceID: 218432495
          Type: IntegratedGpu
        Driver: "MoltenVK"
    DriverInfo: "1.2.6"
     Compliant: true
[....]
Adapter 1:
       Backend: Metal
          Name: "Apple M1 Pro"
      VendorID: 0
      DeviceID: 0
          Type: IntegratedGpu
        Driver: ""
    DriverInfo: ""
     Compliant: true
[....]
```

but if I run the cube example, I'm unable to use it:

```
$ env WGPU_BACKEND=vulkan ./target/release/wgpu-examples cube
[2023-11-23T23:34:37Z INFO  wgpu_examples::framework] Initializing wgpu...
thread 'main' panicked at 'No suitable GPU adapters found on the system!', examples/src/framework.rs:272:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The examples do work correctly with the Metal backend, of course.

I've investigated a bit by tweaking the source code, but it seems that, in `framework.rs`, part of the examples, the Vulkan backend does not exist / is not found / cannot be initialized. I tried several different ways of enumerating / filtering the backend, and it seems not to be there period.

I'm not sure what's the difference in builds between `wgpu-info` and `wgpu-examples`. I would have expected to either both being able to enumerate the Vulkan backend or not, while it seems that one can and the other cannot.

Any suggestion?

**Comment by cwfitzgerald** [comment]
Vulkan on metal is only available if the `vulkan-portability` wgpu feature flag is on. wgpu-info and wgpu-test both turn this on, but the examples don't.

If you run `cargo run --bin wgpu-examples --features vulkan-portability` it should work.

**Reply by rasky** [reply]
Thanks. Maybe this is now becoming more of a Rust question than a wgpu question, but I noticed what follows.
If I use the command line you provided, it does seem to work:

```
$ cargo run --release --bin wgpu-examples --features vulkan-portability cube
    Finished release [optimized] target(s) in 0.15s
     Running `target/release/wgpu-examples cube`
[2023-11-25T12:25:08Z INFO  wgpu_examples::framework] Initializing wgpu...
[2023-11-25T12:25:08Z INFO  wgpu_examples::framework] Using Apple M1 Pro (Vulkan)
```

But if I run directly the binary, it doesn't:

```
$ ./target/release/wgpu-examples cube
[2023-11-25T12:26:38Z INFO  wgpu_examples::framework] Initializing wgpu...
[2023-11-25T12:26:38Z INFO  wgpu_examples::framework] Using Apple M1 Pro (Metal)
```

```
$ env WGPU_BACKEND=vulkan ./target/release/wgpu-examples cube
[2023-11-25T12:26:56Z INFO  wgpu_examples::framework] Initializing wgpu...
thread 'main' panicked at 'No suitable GPU adapters found on the system!', examples/src/framework.rs:273:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I don't know what `cargo run --features` does. I honestly thought it just ran `cargo build --features` and then run the binary, but it doesn't seem that way.

I also rebuilt the whole project with `cargo clean && cargo build --release --features vulkan-portability` in case it changed something but it didn't.
